File: leetcode_ly/c++/List/ReorderList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        if(!head || !head->next || !head->next->next) return;
        ListNode* node = getMiddleNode(head);
        ListNode* next = node->next;
        node->next = NULL;
        ListNode* back = reverse(next);
        ListNode* Head = head;
        while(back){
            ListNode* temp1 = Head->next;
            ListNode* temp2 = back->next;
            Head->next = back;
            back->next = temp1;
            Head = temp1;
            back = temp2;
        }
    }
    
    ListNode* getMiddleNode(ListNode* head){
        ListNode* slow = head;
        ListNode* fast = head;
        while(fast->next && fast->next->next){
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }
    
    ListNode* reverse(ListNode* head){
        if(!head || !head->next) return head;
        ListNode* p = NULL;
        ListNode* q = head;
        ListNode* r = head->next;
        while(r){
            q->next = p;
            p = q;
            q = r;
            r = r->next;
        }
        q->next = p;
        return q;
    }
};

File: leetcode_ly/java/List/ReorderList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void reorderList(ListNode head) {
        if(head == null || head.next == null) return;
        ListNode slow = head;
        ListNode fast = head;
        while(fast.next != null && fast.next.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode temp = slow.next;
        slow.next = null;
        ListNode node = reverse(temp);
        //下面这句话的初始化无意义，是为了保存node当node长度比head长时，将尾指针补上
        ListNode tail = node;
        while(node != null && head != null){
            tail = node;
            ListNode ptr = head.next;
            ListNode btr = node.next;
            head.next = node;
            node.next = ptr;
            head = ptr;
            node = btr;
        }
        if(head == null && node != null) tail.next = node;
    }
    
    public ListNode reverse(ListNode head){
        ListNode p = null;
        ListNode q = head;
        while(head.next != null){
            head = head.next;
            q.next = p;
            p = q;
            q = head;
        }
        head.next = p;
        return head;
    }
}